qcsubmit/results.py

```diff
@@ -0,0 +1,674 @@
+        """
+
+        connectivity_changes = dict(
+            (index, opt_rec.detect_connectivity_changes_heuristic()) for index, opt_rec in enumerate(self.entries)
+        )
+        return connectivity_changes
+
+    def detect_hydrogen_bonds_wbo(self, wbo_threshold: float = 0.05) -> Dict[int, bool]:
+        """
+        Detect hydrogen bonds in the final molecules using the wbo.
+
+        Returns
+        -------
+        hydrogen_bonds :  Dict[int, bool],
+            A dictionary of the optimization entry and a bool representing if an internal hydrogen bond was found.
+            `True` indicates a bond was found.
+            `False` indicates a bond was not found.
+
+        Notes
+        -----
+            You can also query where the hydrogen bond is formed using the `find_hydrogen_bonds_wbo` function on the
+            corresponding entry.
+        """
+        hydrogen_bonds = {}
+        for index, opt_rec in enumerate(self.entries):
+            hbonds = opt_rec.find_hydrogen_bonds_wbo(wbo_threshold)
+            result = bool(hbonds) if hbonds is not None else None
+            hydrogen_bonds[index] = result
+
+        return hydrogen_bonds
+
+    def detect_hydrogen_bonds_heuristic(self) -> Dict[int, bool]:
+        """
+        Detect hydrogen bonds in the final molecule of the trajectory using the Baker-Hubbard rule based method.
+
+        Returns
+        -------
+        hydrogen_bonds : Dict[int, bool],
+            A dictionary of the optimization entry index and bool representing if an internal hydrogen bond was found.
+            `True` indicates a bond was found.
+            `False` indicates a bond was not found.
+
+        Notes
+        -----
+            You can also query which atoms the bond was formed between using the `find_hydrogen_bonds_heuristic` method
+            on the corresponding entry.
+        """
+
+         #TODO
+
+    @property
+    def n_entries(self) -> int:
+        """
+        Get the number of optimization entries for this molecule.
+        """
+
+        return len(self.entries)
+
+
+class OptimizationCollectionResult(BaseModel):
+    """
+    The master collection result which contains many optimizationResults representing the archive collection.
+    """
+
+    method: str
+    basis: str
+    dataset_name: str
+    program: str
+    driver: str
+    scf_properties: List[str]
+    spec_name: str
+    collection: Dict[str, OptimizationResult] = {}
+
+    class Config:
+        arbitrary_types_allowed: bool = True
+        allow_mutation: bool = False
+        json_encoders: Dict[str, Any] = {np.ndarray: lambda v: v.flatten().tolist()}
+
+    @classmethod
+    def from_server(
+        cls,
+        client: ptl.FractalClient,
+        spec_name: str,
+        dataset_name: str,
+        include_trajectory: bool = False,
+        final_molecule_only: Optional[bool] = False,
+        subset: Optional[List[str]] = None,
```

Comment:
It would be nice to have a "only grab the first N records in this dataset" option to avoid `subset = list(opt_ds.data.records.keys())[:100]` that was needed in the example notebook. Not sure this is smart to have in the main class but maybe a util function somewhere, since it may only be useful for testing

Reply:
Good point am not sure how to introduce this though, maybe I could have a prepopulated list of 1000 keys stored in the package for testing?

Reply:
For now really all I need is the ability to grab N random records, which is something that we probably don't want accessible to the user in the main function. So I could just add that

For testing purposes, maybe that pre-populated list of keys that you know the behavior of? So you can know the expected behavior beforehand, usinng the same records each time, but not need to iterate through the entire dataset